### ISSUE TITLE
How to purge & install from scratch?

### ISSUE THREAD

**Issue by brandonmp** [jupyter/notebook#1954]
Ubuntu 16.10

I've been trying to run `jupyter` tonight but keep hitting dead-ends. i've used it successfully on my system before, so this is unexpected.

first, i was receiving this error when starting a new notebook in `jupyter`:

```
[I 22:00:51.384 NotebookApp] [nb_conda_kernels] enabled, 2 kernels found
[I 22:00:51.586 NotebookApp] ✓ nbpresent HTML export ENABLED
[W 22:00:51.586 NotebookApp] ✗ nbpresent PDF export DISABLED: No module named nbbrowserpdf.exporters.pdf
[I 22:00:51.588 NotebookApp] [nb_conda] enabled
[I 22:00:51.627 NotebookApp] [nb_anacondacloud] enabled
[I 22:00:51.628 NotebookApp] Serving notebooks from local directory: /home/bmp/code/wayhome/database-utils
[I 22:00:51.628 NotebookApp] 0 active kernels 
[I 22:00:51.629 NotebookApp] The Jupyter Notebook is running at: http://localhost:8888/
[I 22:00:51.629 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
Created new window in existing browser session.
[E 22:00:52.178 NotebookApp] Unhandled error in API request
    Traceback (most recent call last):
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/notebook/base/handlers.py", line 457, in wrapper
        result = yield gen.maybe_future(method(self, *args, **kwargs))
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/notebook/services/kernelspecs/handlers.py", line 56, in get
        for kernel_name in ksm.find_kernel_specs():
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/nb_conda_kernels/manager.py", line 192, in find_kernel_specs
        kspecs = super(CondaKernelSpecManager, self).find_kernel_specs()
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/jupyter_client/kernelspec.py", line 128, in find_kernel_specs
        for kernel_dir in self.kernel_dirs:
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/traitlets/traitlets.py", line 554, in __get__
        return self.get(obj, cls)
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/traitlets/traitlets.py", line 533, in get
        value = self._validate(obj, dynamic_default())
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/jupyter_client/kernelspec.py", line 114, in _kernel_dirs_default
        from IPython.paths import get_ipython_dir
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/IPython/__init__.py", line 48, in <module>
        from .core.application import Application
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/IPython/core/application.py", line 25, in <module>
        from IPython.core import release, crashhandler
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/IPython/core/crashhandler.py", line 28, in <module>
        from IPython.core import ultratb
      File "/home/bmp/anaconda2/lib/python2.7/site-packages/IPython/core/ultratb.py", line 131, in <module>
        import IPython.utils.colorable as colorable
    AttributeError: 'module' object has no attribute 'utils'
[E 22:00:52.181 NotebookApp] {
      "Accept-Language": "en-US,en;q=0.8", 
      "Accept-Encoding": "gzip, deflate, sdch, br", 
      "X-Requested-With": "XMLHttpRequest", 
      "Host": "localhost:8888", 
      "Accept": "*/*", 
      "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36", 
      "Connection": "keep-alive", 
      "Referer": "http://localhost:8888/tree", 
      "Cookie": "userUUID=5fa79270-5c52-11e6-be10-3b843276dfbb; ajs_anonymous_id=%2224dae7db-3616-4d90-a9c8-d45128d0958b%22; ajs_user_id=null; ajs_group_id=null; _ga=GA1.1.2055710948.1471310596; wh-uid=7ca7f3f4-e818-49bf-bc9d-6862c606c61e; wayhome-cookie={%22visitorID%22:%221e5eb3f2-c39f-417a-bbab-deeb1a4131fa%22}; _hp2_props.3585080487=%7B%22EXP-landingHeadlineExp%22%3A%22this%22%7D; _hp2_id.3585080487=%7B%22userId%22%3A%227693873235831388%22%2C%22pageviewId%22%3A%226746410460421727%22%2C%22sessionId%22%3A%228967741494908047%22%2C%22identity%22%3Anull%2C%22trackerVersion%22%3A%223.0%22%7D; visitor_data={%22visitorID%22:%221b9bcfa4-5939-4502-934a-9acb3c2e0107%22}; grafana_user=admin; grafana_remember=4c42a32af0c39f55951cc62e4426677f5b2f187c918cddf5"
    }
[E 22:00:52.181 NotebookApp] 500 GET /api/kernelspecs (127.0.0.1) 4.74ms referer=http://localhost:8888/tree
```

I tried a few things to fix that, mostly using previously answered questions in this repo/on SO, but ultimately I ended up trying to start fresh, so I:

- deleted and uninstalled `anaconda` 
- uninstalled `python` 
- ensured `ipython` or anything like it was removed via `apt` package mangaer
- downloaded and built-from-source `python 2.7.12`
- installed `pip` and `jupyter` via pip

At that point, I ran `jupyter`, and got: 
```
[I 22:24:11.134 NotebookApp] Serving notebooks from local directory: /home/bmp/code/wayhome/database-utils
[I 22:24:11.134 NotebookApp] 0 active kernels 
[I 22:24:11.134 NotebookApp] The Jupyter Notebook is running at: http://localhost:8888/
[I 22:24:11.134 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
Created new window in existing browser session.
[W 22:24:11.689 NotebookApp] Native kernel (python2) is not available
[W 22:24:11.690 NotebookApp] Native kernel (python2) is not available
[W 22:24:11.694 NotebookApp] Native kernel (python2) is not available
```

So, I ran `pip install ipykernel`, which was met with 100% `Requirements already satisfied`

Running `ipython` from command line:
```
$ ipython
Traceback (most recent call last):
  File "/usr/local/bin/ipython", line 7, in <module>
    from IPython import start_ipython
  File "/usr/local/lib/python2.7/site-packages/IPython/__init__.py", line 48, in <module>
    from .core.application import Application
  File "/usr/local/lib/python2.7/site-packages/IPython/core/application.py", line 25, in <module>
    from IPython.core import release, crashhandler
  File "/usr/local/lib/python2.7/site-packages/IPython/core/crashhandler.py", line 28, in <module>
    from IPython.core import ultratb
  File "/usr/local/lib/python2.7/site-packages/IPython/core/ultratb.py", line 128, in <module>
    from IPython.utils.terminal import get_terminal_size
  File "/usr/local/lib/python2.7/site-packages/IPython/utils/terminal.py", line 22, in <module>
    from backports.shutil_get_terminal_size import get_terminal_size as _get_terminal_size
ImportError: No module named shutil_get_terminal_size
```

On the advise of someone in a GH Issue, I went into `python` and ran:

`>>> from ipykernel.kernelspec import RESOURCES`
which yielded the same error

The only trouble-shooting I can find on this situation is that I might have conflicting `ipython` versions, though I've tried to purge everything. 

Is there a recommended way of  dealing w/ this? Don't want to have to run `jupyter` in a vm or docker :)

**Comment by takluyver** [comment]
The `backports` namespace package seems to cause problems occasionally; we haven't worked out the root cause, but generally finding and removing all `backports.something` packages, then reinstalling those you need, makes it work.

The idea of a namespace package is that the different packages in it can be installed in different places on your system. But if one of those gets it wrong, it can hide other packages so import can't find them at all.

**Comment by brandonmp** [comment]
Okay, yeah I figured it was some packages hiding somewhere, based on previous things you've written on the topic (thanks, btw).  

I don't know if there's a causal link here, but I ran:

`sudo -H pip freeze | sudo -H xargs pip uninstall -y` 

to remove all packages/modules, and after re-installing `jupyter` and a few other packages, `jupyter` appeared to run fine. 


A shame none of the full-purge steps I tried managed to get _all_ modules, but glad it works now. thanks @takluyver !
